### PR TITLE
Automate version number handling via versioneer

### PR DESCRIPTION
**Pull request by grlee77**
In #170 @hgomersall requested improvements related to simplifying the release process.

Toward that goal, I made a PR to use [versioneer](https://github.com/warner/python-versioneer) so that it is no longer necessary to go in and manually update version strings and release flags in `setup.py`.  I would be interested if other think this is a good fit for this project.

Basically `versioneer` creates a reasonable PEP440 compliant version string based on the most recent git tag as described at the README linked above.  The `versioneer.py` file added here was autogenerated via the `versioneer` command line script based on the contents of the newly added `setup.cfg`.

For example, after building from this PR I get:
```python
import pyfftw
print("Version = {}".format(pyfftw.__version__))
```
```
Version = v0.10.4+64.g3b217ce
```
indicating that we are 64 commits ahead of the most recent tag.

Going this route also makes `setup.py` itself considerably simpler.

### CODE REVIEW

**Comment by hgomersall** [comment]
This looks like an obvious thing to do and versioneer seems like it will do the job well :)

Do we need to include `versioneer.py`? It doesn't seem an unreasonable dependency for the user to install given that its in PyPI.

**Comment by hgomersall** [comment]
On a parallel issue, I've been thinking about something like [releaser](https://github.com/nandoflorestan/releaser) for managing the release as well... (I don't know how much overlap of conflict there is between that and versioneer).

**Comment by grlee77** [comment]
I have seen other projects leave `versioneer.py` there so that no dependency is added (e.g. see https://github.com/dask/dask).  The build may succeed without it as long as the user has versioneer, but I haven't tried it.

As far as `releaser`, I am not familiar with it and haven't looked at it in any detail, but just examining the primary script, I think if you are using versioneer you would comment out the following two lines:
https://github.com/nandoflorestan/releaser/blob/3be066161711aad3f89b52d1cee91ddafca71d04/release_new_version.py#L43-L44

**Comment by hgomersall** [comment]
What does `versioneer.py` do and what does `_version.py` do? I'm a little confused...

Other than that, in principle it looks great. Should this have any issues or should it be seamless?

**Comment by grlee77** [comment]
I double checked that even if the versioneer module `versioneer` is installed on the user's machine, `versioneer.py` is required to be present in the same folder as `setup.py`.  The module is mainly just an installer for `versioneer.py`.

>> What does versioneer.py do and what does _version.py do? I'm a little confused...

I think the design is a little confusing, but this small section of the readme explains the different purposes of the two:
https://github.com/warner/python-versioneer#theory-of-operation

Basically during the initial install of versioneer the `do_setup` method from `versioneer.py` generated the `_version.py` file based on values in `site.cfg`, added it to MANIFEST.in, etc.  The methods related to use in `setup.py` are in `versioneer.py` while the `_version.py` has the functions for querying versions from `git`.

For some types of builds (e.g. source distribution) `versioneer.py` replaces `_version.py` with a more minimal file as the version is fixed for that particular source release.  I tried it and the version that ends up in a source distribution looks something like:
```python
# This file was generated by 'versioneer.py' (0.18) from
# revision-control system data, or from the parent directory name of an
# unpacked source archive. Distribution tarballs contain a pre-generated copy
# of this file.

import json

version_json = '''
{
 "date": "2017-05-10T18:19:06-0400",
 "dirty": false,
 "error": null,
 "full-revisionid": "e276406ca3954dc7e81bf030a2672fac0fc0ac43",
 "version": "v0.10.4+65.ge276406"
}
'''  # END VERSION_JSON


def get_versions():
    return json.loads(version_json)
```

File: setup.cfg [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[versioneer]
+VCS = git
+style = pep440
+versionfile_source = pyfftw/_version.py
+versionfile_build = pyfftw/_version.py
+tag_prefix =
```
Comment (jakirkham) [inline setup.cfg]: This probably needs to be `v` as the tags here seem to be also prefixing with `v`. Also I think one needs to rerun `versioneer install` after changing this file.

**Comment by jakirkham** [comment]
> Do we need to include `versioneer.py`? It doesn't seem an unreasonable dependency for the user to install given that its in PyPI.

This gets included in `sdist`s or other release products that may be sent to PyPI as it is in the `MANIFEST.in` file. So @grlee77 is correct that there is no dependency. The file is merely included in any package the user would download.

**Comment by jakirkham** [comment]
FWIW am a big fan of `versioneer` and use it religiously in all my projects. We even use it at conda-forge in `conda-smithy`. It's also used in many projects over at Continuum as well. As it is dependency free (all relevant code lives in our codebase) and in the Public Domain (CC0-1.0), it remains the easiest versioning tool for Python I have seen.

While it may seem to be very complicated, basically `versioneer` parses `git describe` and turns it into a [PEP 440]( https://www.python.org/dev/peps/pep-0440/ ) compliant version string. There is a bit of fiddling to get the result from `git describe` to match PEP 440. However, as there is not always a git repo involved (e.g. an `sdist`), it also needs some other mechanisms for tracking this information once such products are created. Also one can be building a dev version, so it has to handle that somehow. Hence the rest of the code is handling these transformations and edge cases.

All in all it is pretty slick. Hopefully that gives you some appreciation of why.

**Comment by hgomersall** [comment]
I'm convinced!

@jakirkham It seems that `versioneer.py` in this case is part of the git repository. I don't see how that corresponds to it being pulled in with sdist...

File: MANIFEST.in [hunk 2]
```diff
@@ -12,3 +12,5 @@ include index.rst
 recursive-include pyfftw *.rst
 recursive-include sphinx *.rst
 recursive-include include *.h
+include versioneer.py
+include pyfftw/_version.py
```
Comment (jakirkham) [inline MANIFEST.in]: These lines pull `versioneer.py` and `pyfftw/_version.py` into the `sdist`.

**Comment by jakirkham** [comment]
> I'm convinced!

Sorry that turned into a sales pitch. 😄

File: setup.cfg (hunk 1 above)
Comment (grlee77) [inline setup.cfg]: Good catch.  I have updated the PR accordingly.  `pyfftw.__version__` now reports `'0.10.4+66.g8e72f05'` in this branch

**Comment by jakirkham** [comment]
Sorry missed responding to this for some reason, @hgomersall.

> It seems that `versioneer.py` in this case is part of the git repository. I don't see how that corresponds to it being pulled in with sdist...

It's listed in the `MANIFEST.in`. Hence it gets added to the `sdist` that way.

**Comment by jakirkham** [comment]
Is it ok if I merge this, @hgomersall? Or were there other things you wanted to discuss?

**Comment by hgomersall** [comment]
I think we can merge this once the conflicts are resolved (in light of #177 being merged).

**Pull request merged**